### DISCUSSION TITLE
使用Cygwin编译源码失败，求教。。。

### DISCUSSION THREAD

**Discussion by yutf** [nanomq/nanomq#1711] [Q&A]
$ cmake -G Ninja -DBUILD_SHARED_LIBS=ON ..
CMake Deprecation Warning at CMakeLists.txt:8 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.

-- The C compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning (dev) at CMakeLists.txt:138 (message):
  WARNING: This platform may not be supported: CYGWIN
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at CMakeLists.txt:139 (message):
This warning is for project developers.  Use -Wno-dev to suppress it.

-- NanoMQ versions --
-- Configuring for NanoMQ version 0.21.6
-- Build NanoMQ with ACL support --
-- MODE [Release] --
NNG_TESTS=ON
NNG_TOOLS=ON
-- Configuring for NNG version 1.7.2
CMake Warning (dev) at nng/CMakeLists.txt:283 (message):
  WARNING: This platform may not be supported: CYGWIN
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at nng/CMakeLists.txt:284 (message):
  Please consider opening an issue at https://github.com/nanomq/NanoNNG
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Looking for strlcpy
-- Looking for strlcpy - found
-- Looking for strnlen
-- Looking for strnlen - found
-- Looking for strcasecmp
-- Looking for strcasecmp - found
-- Looking for strncasecmp
-- Looking for strncasecmp - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Looking for lockf
-- Looking for lockf - found
-- Looking for flock
-- Looking for flock - found
-- Looking for getrandom
-- Looking for getrandom - found
-- Looking for arc4random_buf
-- Looking for arc4random_buf - found
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - found
-- Looking for sem_wait in pthread
-- Looking for sem_wait in pthread - found
-- Looking for pthread_atfork in pthread
-- Looking for pthread_atfork in pthread - found
-- Looking for pthread_set_name_np in pthread
-- Looking for pthread_set_name_np in pthread - not found
-- Looking for pthread_setname_np in pthread
-- Looking for pthread_setname_np in pthread - found
-- Looking for gethostbyname in nsl
-- Looking for gethostbyname in nsl - not found
-- Looking for socket in socket
-- Looking for socket in socket - not found
-- Looking for __atomic_load_1 in atomic
-- Looking for __atomic_load_1 in atomic - found
-- Looking for AF_UNIX
-- Looking for AF_UNIX - found
-- Looking for backtrace_symbols_fd
-- Looking for backtrace_symbols_fd - not found
-- Performing Test NNG_HAVE_MSG_CONTROL
-- Performing Test NNG_HAVE_MSG_CONTROL - Success
-- Looking for eventfd
-- Looking for eventfd - not found
-- Looking for kqueue
-- Looking for kqueue - not found
-- Looking for port_create
-- Looking for port_create - not found
-- Looking for epoll_create
-- Looking for epoll_create - not found
-- Looking for epoll_create1
-- Looking for epoll_create1 - not found
-- Looking for getpeereid
-- Looking for getpeereid - found
-- Looking for SO_PEERCRED
-- Looking for SO_PEERCRED - found
-- Performing Test NNG_HAVE_SOCKPEERCRED
-- Performing Test NNG_HAVE_SOCKPEERCRED - Failed
-- Looking for LOCAL_PEERCRED
-- Looking for LOCAL_PEERCRED - not found
-- Looking for LOCAL_PEERPID
-- Looking for LOCAL_PEERPID - not found
-- Looking for getpeerucred
-- Looking for getpeerucred - not found
-- Looking for atomic_flag_test_and_set
-- Looking for atomic_flag_test_and_set - found
-- Looking for socketpair
-- Looking for socketpair - found
-- Looking for AF_INET6
-- Looking for AF_INET6 - found
 Check MQTT_QUIC_CLIENT support:
 Check MQTT_QUIC_TRANSPORT support:
-- Looking for nl_langinfo
-- Looking for nl_langinfo - found
-- Test zt disabled (unconfigured)
CMake Deprecation Warning at nanomq/CMakeLists.txt:7 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.

CMake Deprecation Warning at nanomq_cli/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.

Install prefix: /usr/local
-- Configuring done (19.9s)
-- Generating done (0.8s)
-- Build files have been written to: /home/Administrator/nanomq/build

Administrator@hsdev ~/nanomq/build
$ ninja
[32/544] Building C object nng/CMakeFiles/nng.dir/src/core/url.c.o
In file included from /home/Administrator/nanomq/nng/src/core/url.c:13:
/home/Administrator/nanomq/nng/src/core/url.c: In function ‘nni_url_decode’:
/home/Administrator/nanomq/nng/src/core/url.c:104:42: warning: array subscript has type ‘char’ [-Wchar-subscripts]
  104 |                         if ((!isxdigit(in[0])) || (!isxdigit(in[1]))) {
      |                                        ~~^~~
/home/Administrator/nanomq/nng/src/core/url.c:104:64: warning: array subscript has type ‘char’ [-Wchar-subscripts]
  104 |                         if ((!isxdigit(in[0])) || (!isxdigit(in[1]))) {
      |                                                              ~~^~~
/home/Administrator/nanomq/nng/src/core/url.c: In function ‘url_canonify_uri’:
/home/Administrator/nanomq/nng/src/core/url.c:138:43: warning: array subscript has type ‘char’ [-Wchar-subscripts]
  138 |                         if ((!isxdigit(out[src + 1])) ||
      |                                        ~~~^~~~~~~~~
/home/Administrator/nanomq/nng/src/core/url.c:139:43: warning: array subscript has type ‘char’ [-Wchar-subscripts]
  139 |                             (!isxdigit(out[src + 2]))) {
      |                                        ~~~^~~~~~~~~
[45/544] Building C object nng/CMakeFiles/nng.dir/src/platform/posix/posix_ipclisten.c.o
/home/Administrator/nanomq/nng/src/platform/posix/posix_ipclisten.c: In function ‘ipc_listener_listen’:
/home/Administrator/nanomq/nng/src/platform/posix/posix_ipclisten.c:313:56: warning: comparison of integer expressions of different signedness: ‘socklen_t’ {aka ‘int’} and ‘long unsigned int’ [-Wsign-compare]
  313 |         if ((len = nni_posix_nn2sockaddr(&ss, &l->sa)) < sizeof(sa_family_t)) {
      |                                                        ^
[46/544] Building C object nng/CMakeFiles/nng.dir/src/platform/posix/posix_thread.c.o
FAILED: nng/CMakeFiles/nng.dir/src/platform/posix/posix_thread.c.o
/usr/bin/cc -DACL_SUPP -DENABLE_LOG -DNANO_PLATFORM_POSIX -DNNG_ENABLE_COMPAT -DNNG_ENABLE_IPV6 -DNNG_ENABLE_STATS -DNNG_HAVE_ARC4RANDOM=1 -DNNG_HAVE_BUS0 -DNNG_HAVE_CLOCK_GETTIME=1 -DNNG_HAVE_FLOCK=1 -DNNG_HAVE_GETPEEREID=1 -DNNG_HAVE_GETRANDOM=1 -DNNG_HAVE_INET6=1 -DNNG_HAVE_LANGINFO=1 -DNNG_HAVE_LIBATOMIC=1 -DNNG_HAVE_LOCKF=1 -DNNG_HAVE_MQTT_BROKER -DNNG_HAVE_MQTT_CLIENT -DNNG_HAVE_MSG_CONTROL=1 -DNNG_HAVE_PAIR0 -DNNG_HAVE_PAIR1 -DNNG_HAVE_PTHREAD_ATFORK_PTHREAD=1 -DNNG_HAVE_PTHREAD_SETNAME_NP=1 -DNNG_HAVE_PUB0 -DNNG_HAVE_PULL0 -DNNG_HAVE_PUSH0 -DNNG_HAVE_REP0 -DNNG_HAVE_REQ0 -DNNG_HAVE_RESPONDENT0 -DNNG_HAVE_SEMAPHORE_PTHREAD=1 -DNNG_HAVE_SOCKETPAIR=1 -DNNG_HAVE_SOPEERCRED=1 -DNNG_HAVE_STDATOMIC=1 -DNNG_HAVE_STRCASECMP=1 -DNNG_HAVE_STRLCPY=1 -DNNG_HAVE_STRNCASECMP=1 -DNNG_HAVE_STRNLEN=1 -DNNG_HAVE_SUB0 -DNNG_HAVE_SURVEYOR0 -DNNG_HAVE_UNIX_SOCKETS=1 -DNNG_MAX_EXPIRE_THREADS=8 -DNNG_MAX_POLLER_THREADS=8 -DNNG_MAX_TASKQ_THREADS=16 -DNNG_PLATFORM_POSIX -DNNG_PRIVATE -DNNG_RESOLV_CONCURRENCY=4 -DNNG_SHARED_LIB -DNNG_SUPP_HTTP -DNNG_TRANSPORT_FDC -DNNG_TRANSPORT_INPROC -DNNG_TRANSPORT_IPC -DNNG_TRANSPORT_MQTT_BROKER_TCP -DNNG_TRANSPORT_MQTT_BROKER_TLS -DNNG_TRANSPORT_MQTT_BROKER_WS -DNNG_TRANSPORT_MQTT_TCP -DNNG_TRANSPORT_MQTT_TLS -DNNG_TRANSPORT_TCP -DNNG_TRANSPORT_TLS -DNNG_TRANSPORT_WS -DSUPP_CLIENT -DSUPP_NANO_LIB -DSUPP_NNG_PROXY -DSUPP_SYSLOG -D_GNU_SOURCE -D_POSIX_PTHREAD_SEMANTICS -D_REENTRANT -D_THREAD_SAFE -Dnng_EXPORTS -I/home/Administrator/nanomq/nng/src -I/home/Administrator/nanomq/nng/include -I/home/Administrator/nanomq/nanolib/include -I/home/Administrator/nanomq/nanomq/include -Wall -Wextra -fno-omit-frame-pointer   -std=gnu99 -MD -MT nng/CMakeFiles/nng.dir/src/platform/posix/posix_thread.c.o -MF nng/CMakeFiles/nng.dir/src/platform/posix/posix_thread.c.o.d -o nng/CMakeFiles/nng.dir/src/platform/posix/posix_thread.c.o -c /home/Administrator/nanomq/nng/src/platform/posix/posix_thread.c
In file included from /home/Administrator/nanomq/nng/include/nng/supplemental/nanolib/log.h:12,
                 from /home/Administrator/nanomq/nng/src/core/nng_impl.h:15,
                 from /home/Administrator/nanomq/nng/src/platform/posix/posix_thread.c:13:
/home/Administrator/nanomq/nng/src/platform/posix/posix_thread.c: In function ‘nni_plat_init’:
/home/Administrator/nanomq/nng/src/platform/posix/posix_config.h:71:25: error: missing binary operator before token "4"
   71 | #define NNG_USE_CLOCKID CLOCK_MONOTONIC
      |                         ^~~~~~~~~~~~~~~
/home/Administrator/nanomq/nng/src/platform/posix/posix_thread.c:368:39: note: in expansion of macro ‘NNG_USE_CLOCKID’
  368 | #if !defined(NNG_USE_GETTIMEOFDAY) && NNG_USE_CLOCKID != CLOCK_REALTIME
      |                                       ^~~~~~~~~~~~~~~
[71/544] Building C object nng/CMakeFiles/nng.dir/src/sp/protocol/mqtt/nmq_mqtt.c.o
ninja: build stopped: subcommand failed.

这个是编译环境的问题吗？有没有解决方法？本人对C\C++技术栈了解不深，尝试了很久很久都没有成功~~~~~

**Comment by StargazerWayne** [comment]
如果你的需求只是在windows上编译nanomq，nanomq其实支持直接在windows上编译安装，详细过程可以参见[教程](https://www.emqx.com/zh/blog/install-mqtt-broker-on-windows)。
